File: 274-h-index/solution.cpp
#include <vector>
#include <algorithm>

class Solution
{
public:
  int hIndex(std::vector<int> &citations)
  {
    std::vector<int> frequencies(citations.size() + 1, 0);
    for (const size_t num_citations : citations)
    {
      ++frequencies[std::min(num_citations, frequencies.size() - 1)];
    }

    size_t extra = 0;
    for (size_t num_citations = frequencies.size() - 1; num_citations >= 0; --num_citations)
    {
      const auto num_papers = frequencies[num_citations];
      if (num_papers + extra >= num_citations)
      {
        return num_citations;
      }

      extra += num_papers;
    }

    return 0;
  }
};

File: 274-h-index/solution.py
from typing import List


class Solution:
  def hIndex(self, citations: List[int]) -> int:
    frequencies = [0 for _ in range(len(citations) + 1)]
    for num_citations in citations:
      frequencies[min(num_citations, len(frequencies) - 1)] += 1

    extra = 0
    for num_citations in reversed(range(len(frequencies))):
      num_papers = frequencies[num_citations]
      if num_papers + extra >= num_citations:
        return num_citations

      extra += num_papers

    return 0
